File: MS-MPI/Assignment3/task1.cpp
#include <mpi.h>
#include <iostream>
#include <vector>
using namespace std;

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);

    int world_rank, world_size;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank); 
    MPI_Comm_size(MPI_COMM_WORLD, &world_size); // Get total number of ps

    vector<int> array = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int array_size = array.size();
    int chunk_size = array_size / world_size;
    int start_index = world_rank * chunk_size;
    int end_index = (world_rank == world_size - 1) ? array_size : start_index + chunk_size;

    double partial_sum = 0;
    for (int i = start_index; i < end_index; i++) {
        partial_sum += array[i];
    }

    cout << "Number of Processes available is " << world_size <<"\n";
    cout << "\nSUM - Process " << world_rank << ": \n\t My contribution to the sum is " << partial_sum << "\n";
    double total_sum = 0;
    MPI_Reduce(&partial_sum, &total_sum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
    if (world_rank == 0) {
        cout << "SUM - Master process: \n\t The total sum is " << total_sum << "\n";
        cout << "SUM - Master process: \n\t Normal end of execution.\n";
    }
    MPI_Finalize();
    return 0;
}

File: MS-MPI/pyy/task_1.py
from mpi4py import MPI

comm = MPI.COMM_WORLD
world_rank = comm.Get_rank() 
world_size = comm.Get_size() 

array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
array_size = len(array)
chunk_size = array_size // world_size
start_index = world_rank * chunk_size
end_index = array_size if world_rank == world_size - 1 else start_index + chunk_size

partial_sum = sum(array[start_index:end_index])

print(f"Sum - Process {world_rank}:\n My contribution to the sum is {partial_sum}")


total_sum = comm.reduce(partial_sum, op=MPI.SUM, root=0)

if world_rank == 0:
    print(f"The total sum calculated by the master process is {total_sum}")
    print("Sum-Master: \nNormal end of execution.")
